lib/view_component/slotable_v2.rb

```diff
@@ -0,0 +1,219 @@
+# frozen_string_literal: true
+
+require "active_support/concern"
+
+require "view_component/slot"
+
+module ViewComponent
+  module Slotable
+    ##
+    # Version 2 of the Slots API
+    module V2
+      extend ActiveSupport::Concern
+
+      # Setup component slot state
+      included do
+        # Hash of registered Slots
+        class_attribute :registered_slots
+        self.registered_slots = {}
+      end
+
+      class_methods do
+        ##
+        # Registers a slot on the component.
+        #
+        # = Example
+        #
+        #   renders_one :header do
+        #     def initialize(classes:)
+        #       @name = name
+        #     end
+        #   end
+        #
+        # = Rendering slot content
+        #
+        # The component's sidecar template can access the slot by calling a
+        # helper method with the same name as the slot (pluralized if the slot
+        # is a collection).
+        #
+        #   <h1>
+        #     <%= header %>
+        #   </h1>
+        #
+        # = Setting slot content
+        #
+        # Renderers of the component can set the content of a slot by calling a
+        # helper method with the same name as the slot. For collection
+        # components, the method can be called multiple times to append to the
+        # slot.
+        #
+        #   <%= render_inline(MyComponent.new) do |component| %>
+        #     <%= component.header(classes: "Foo") do %>
+        #       <p>Bar</p>
+        #     <% end %>
+        #   <% end %>
+        def renders_one(slot_name, &block)
```

Comment:
The new API accepts a block to achieve the flexibility of defining a class, but I wonder if we could achieve the same flexibility with a different approach. The existing block approach also ends up causing consumers of the `renders_*` API to define initializers for slots if they want to accept arguments. They also have to  expose those arguments somehow (likely via methods or `attr_reader`). This means slots still end up being verbose to implement.

I have a few ideas that I'm kicking around, but I wanted to get thoughts on this idea:

* `renders_*` methods accept a labmda and automatically expose any defined instance variables as method names
* If the return value of `renders_*` return a component, we render it

The first example would look something like:

```ruby
renders_many :items, -> (classes:) { @classes = classes } # callable via `items.classes` in the template
```

and component usage could look like this:

```ruby
renders_many :items, -> (item, classes:) { MyProductComponent.new(item, classes: classes) }
```

I think this gives us a few benefits:

* Simple slots are now more succinct, and can more easily accept arguments.
* More complex slots are now easier to create by using components
* Any existing component can now become a slot in another component
* Common slot behavior can be encapsulated in components

Reply:
Building on that thought, I think we could also have a shortcut syntax for the second example.

`renders_many :items, MyProductComponent`

and "content area slots" like `renders_many :items` would still work with this API.